sensors/weather/sensorhub-driver-kestrel/src/main/java/org/sensorhub/impl/sensor/kestrel/Kestrel.java

```diff
@@ -0,0 +1,378 @@
+                }
+            });
+
+            IGattService enviroService = null;
+            for (IGattService service : gatt.getServices()) {
+                UUID uuid = service.getType();
+                if (uuid.equals(ENVIRONMENTAL_SERVICE)) {
+                    enviroService = service;
+                }
+            }
+
+
+            for (IGattCharacteristic ch : enviroService.getCharacteristics()) {
+                if (ch.getType().equals(SENSOR_MEASUREMENTS_CHAR)) {
+                    sensorMeasurements = ch;
+                    if (sensorMeasurements != null)
+                        enableNotification(gatt, sensorMeasurements);
+                } else if (ch.getType().equals(DERIVED_MEASUREMENTS_1_CHAR)) {
+                    derivedMeasurementsOne = ch;
+                    if (derivedMeasurementsOne != null)
+                        enableNotification(gatt, derivedMeasurementsOne);
+                } else if (ch.getType().equals(DERIVED_MEASUREMENTS_2_CHAR)) {
+                    derivedMeasurementsTwo = ch;
+                    if (derivedMeasurementsTwo != null)
+                        enableNotification(gatt, derivedMeasurementsTwo);
+                } else if (ch.getType().equals(DERIVED_MEASUREMENTS_3_CHAR)) {
+                    derivedMeasurementsThree = ch;
+                    if (derivedMeasurementsThree != null)
+                        enableNotification(gatt, derivedMeasurementsThree);
+                } else if (ch.getType().equals(DERIVED_MEASUREMENTS_4_CHAR)) {
+                    derivedMeasurementsFour = ch;
+                    if (derivedMeasurementsFour != null)
+                        enableNotification(gatt, derivedMeasurementsFour);
+                }
+            }
+        }
+
+
+        @Override
+        public void onDisconnected(IGattClient gatt, int status)
+        {
+            notifyConnectionStatus(false, "Kestrel Weather Sensor (" + config.deviceAddress + ")");
+
+        }
+
+        @Override
+        public void onCharacteristicChanged(IGattClient gatt, IGattField characteristic)
+        {
+            byte[] val = characteristic.getValue().array();
+            handleCharacteristicBytes(characteristic.getType(), val);
+        }
+
+        public void onCharacteristicRead(IGattClient gatt, IGattField characteristic, int status)
+        {
+            logger.debug("Characteristic Read: {}", new String(characteristic.getValue().array()));
+        }
+
+        private void enableNotification(IGattClient gatt, IGattCharacteristic ch) {
+            gatt.setCharacteristicNotification(ch, true);
+
+            IGattDescriptor cccd = ch.getDescriptors().get(CLIENT_CONFIG);
+            if (cccd != null) {
+                cccd.setValue(ByteBuffer.wrap(IGattDescriptor.ENABLE_NOTIFICATION_VALUE));
+                descriptorWriteQueue.add(cccd);
```

Comment:
Why enqueue the IGattDescriptor object if just below you will write the descriptor? Read and write in this case seem to be synchronous and handled by the same thread...

I imagine a queue would be necessary for asynchronous reads and writes in which case your queue is not thread safe.